Support PEP 600 tags

mattip opened the pull request:
xref gh-280. This adds a whole bunch of PEP 600 tags (my python reports `os.confstr('CS_GNU_LIBC_VERSION') of 2.27): **EDIT**: after reversing the order in f432bcd93f4
```
>>> from packaging import tags
>>> print ('\n'.join([str(t) for t in tags.sys_tags()]))
cp37-cp37m-manylinux_2_27_x86_64
cp37-cp37m-manylinux_2_26_x86_64
cp37-cp37m-manylinux_2_25_x86_64
cp37-cp37m-manylinux_2_24_x86_64
cp37-cp37m-manylinux_2_23_x86_64
cp37-cp37m-manylinux_2_22_x86_64
cp37-cp37m-manylinux_2_21_x86_64
cp37-cp37m-manylinux_2_20_x86_64
cp37-cp37m-manylinux_2_19_x86_64
cp37-cp37m-manylinux_2_18_x86_64
cp37-cp37m-manylinux_2_17_x86_64
cp37-cp37m-manylinux2014_x86_64
cp37-cp37m-manylinux_2_16_x86_64
cp37-cp37m-manylinux_2_15_x86_64
cp37-cp37m-manylinux_2_14_x86_64
cp37-cp37m-manylinux_2_13_x86_64
cp37-cp37m-manylinux_2_12_x86_64
cp37-cp37m-manylinux2010_x86_64
cp37-cp37m-manylinux_2_11_x86_64
cp37-cp37m-manylinux_2_10_x86_64
cp37-cp37m-manylinux_2_9_x86_64
cp37-cp37m-manylinux_2_8_x86_64
cp37-cp37m-manylinux_2_7_x86_64
cp37-cp37m-manylinux_2_6_x86_64
cp37-cp37m-manylinux_2_5_x86_64
cp37-cp37m-manylinux1_x86_64
cp37-cp37m-linux_x86_64
cp37-abi3-manylinux_2_27_x86_64
...
```

mattip commented on packaging/tags.py:
```diff
@@ -652,11 +652,28 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+                # glibc in manylinux2014 is 2.5
+                min_minor = 17
+            for glibc_minor in range(min_minor, 50):
+                # PEP 600 - up to glibc 2.50 (as of April 2020 last is 2.31)
```
50 is rather arbitrary, any thoughts of a better stopping mechanism? Would parsing `_glibc_version_string()` be better?

mattip commented:
It seems I need to mock some failures to get coverage to 100%

brettcannon commented:
Thanks for taking a stab at this! I haven't looked at the code, but a tweak we might want to make to the output is to emit the `manylinux1`, `manylinux2010`, and `manylinux2014` tags **after** [the matching perennial manylinux tag](https://www.python.org/dev/peps/pep-0600/#legacy-manylinux-tags). We also want to emit the tags in priority order, so newer glibc versions should come first.

mattip commented:
Ahh, so the correct order, assuming precedence for the perennial tags, would be this?
```
cp37-cp37m-manylinux_2_27_x86_64
cp37-cp37m-manylinux_2_26_x86_64
cp37-cp37m-manylinux_2_25_x86_64
cp37-cp37m-manylinux_2_24_x86_64
cp37-cp37m-manylinux_2_23_x86_64
cp37-cp37m-manylinux_2_22_x86_64
cp37-cp37m-manylinux_2_21_x86_64
cp37-cp37m-manylinux_2_20_x86_64
cp37-cp37m-manylinux_2_19_x86_64
cp37-cp37m-manylinux_2_18_x86_64
cp37-cp37m-manylinux_2_17_x86_64
cp37-cp37m-manylinux2014_x86_64
cp37-cp37m-manylinux_2_16_x86_64
cp37-cp37m-manylinux_2_15_x86_64
cp37-cp37m-manylinux_2_14_x86_64
cp37-cp37m-manylinux_2_13_x86_64
cp37-cp37m-manylinux_2_12_x86_64
cp37-cp37m-manylinux2010_x86_64
cp37-cp37m-manylinux_2_11_x86_64
cp37-cp37m-manylinux_2_10_x86_64
cp37-cp37m-manylinux_2_9_x86_64
cp37-cp37m-manylinux_2_8_x86_64
cp37-cp37m-manylinux_2_7_x86_64
cp37-cp37m-manylinux_2_6_x86_64
cp37-cp37m-manylinux_2_5_x86_64
cp37-cp37m-manylinux1_x86_64
cp37-cp37m-linux_x86_64
```

brettcannon commented:
@mattip yep, exactly! Basically the way to think about the order is if I got every wheel file name for a specific version, I should be able to iterate through what `sys_tags()` returns and as soon as I have a tag match with a file I stop looking because I found the best possible match for my environment.

This is also why everything in `packaging.tags` is a generator: so you can short-circuit the search and potentially avoid costly environment checks.

mattip commented:
I reversed the tags and added tests for lines that weren't in the coverage report. However coverage is now saying the critical part of the code that I added is not covered. How can that be when the tests fail if I remove it?

mattip commented:
Any thoughts? CI is passing, even though travis thinks it is still pending.

brettcannon commented:
Sorry, I've personally been swamped lately on top of vacation, so I just haven't had time to review the PR yet.

brettcannon commented on packaging/tags.py:
```diff
@@ -650,28 +662,40 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+            )
+            if manylinux_compat:
+                yield linux.replace("linux", plat)
+            # Handle the manylinux1, manylinux2010, manylinux2014 tags
```
```suggestion
            # Handle the manylinux1, manylinux2010, manylinux2014 tags.
```

mattip replied:
ok

brettcannon commented on packaging/tags.py:
```diff
@@ -650,28 +662,40 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+                    plat, (2, glibc_minor)
+                )
+                if manylinux_compat:
+                    yield linux.replace("linux", plat)
```
Instead of doing the replacing, maybe the map should just hard-code all possible supported architectures since there aren't that many.

mattip replied:
It would only save one instance of `replace`, the PEP 600 tags do not use the map.

mattip commented on packaging/tags.py:
```diff
@@ -509,11 +516,8 @@ def _check_glibc_version(version_str, required_major, minimum_minor):
-        int(m.group("major")) == required_major
-        and int(m.group("minor")) >= minimum_minor
-    )
+        return (-1, -1)
```
fixing

mattip commented on packaging/tags.py:
```diff
@@ -496,7 +496,14 @@ def _glibc_version_string_ctypes():
-    # Parse string and check against requested version.
+    # Check against requested version.
+    major, minor = _parse_glibc_version(version_str)
+    return major == required_major and minor >= minimum_minor
```
fixing

mattip commented on packaging/tags.py:
```diff
@@ -650,28 +662,40 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
-    manylinux_support = []
     _, arch = linux.split("_", 1)
     if _have_compatible_manylinux_abi(arch):
+        map_glibc_to_manylinux = {}
```
ok

mattip commented on packaging/tags.py:
```diff
@@ -650,28 +662,40 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+                map_glibc_to_manylinux[(2, 5)] = "manylinux1"
+                # glibc in manylinux1 is 2.5
+                min_minor = 4
+        cur_glibc = _get_glibc_version()
```
ok

mattip commented on packaging/tags.py:
```diff
@@ -650,28 +662,40 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+        cur_glibc = _get_glibc_version()
+        manylinux_compat = False
+        for glibc_minor in range(cur_glibc[1], min_minor, -1):
+            plat = "manylinux_2_{}".format(glibc_minor)
```
ok

mattip commented on packaging/tags.py:
```diff
@@ -650,28 +662,40 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
+                map_glibc_to_manylinux[(2, 12)] = "manylinux2010"
+                # CentOS 5 w/ glibc 2.5 (PEP 513)
+                map_glibc_to_manylinux[(2, 5)] = "manylinux1"
+                # glibc in manylinux1 is 2.5
+                min_minor = 4
```
ok

mattip commented:
Tests are passing after fixes as per the review. The code should support newer major versions of glibc, but there still is a logic failure. We assume `sys_tags` can emit the entire set of valid tags as listed in the comments above. However, if the platform has moved to `glibc3.2`, what is the highest `Y` for `manylinux_2_Y`? We have no way of knowing what the "last `glibc2.Y`" is. I could just "assume" an arbitrary large number (99?) and a future PR could adapt that to the situation when it arises?

brettcannon commented:
Huh, good point about the cap on glibc 2's minor number if glibc 3 ever becomes a thing. Maybe we can't generalize this for glibc 3 without hard-coding that? I guess the best fall-back is to go with your suggestion, Matti, and just assume `99` as an upper-bound and if the day comes that glibc 3 exists we can hard-code it going forward. That way we somewhat future-proof for people who don't upgrade while not making it ridiculous for us forever.

mattip commented:
Added a `_LAST_GLIBC_MINOR` default dictionary that defaults to 50, and generalized the looping to handle maximum versions of each glibc major version (including using (2, 5) or (2,17) as the last one, depending on architecture). Probably overkill, but should not cause failures. Of course this assumes glibc will make good on their backward-compatibility promise.

pradyunsg commented on packaging/tags.py:
```diff
@@ -57,6 +58,24 @@
+# be 50 for testing. Once this actually happens, update the dictionary
+# with the actual value.
+_LAST_GLIBC_MINOR = collections.defaultdict(lambda: 50)  # type: Dict[int, int]
+version = collections.namedtuple("Version", ["major", "minor"])
```
```suggestion
glibcVersion = collections.namedtuple("glibcVersion", ["major", "minor"])
```

To avoid any confusion with `packaging.version`: https://packaging.pypa.io/en/latest/version/.

mattip replied:
ok

mattip commented on tests/test_tags.py:
```diff
@@ -509,18 +546,50 @@ def test_linux_platforms_manylinux2014_i386_abi(self, monkeypatch):
         )
         platforms = list(tags._linux_platforms(is_32bit=True))
         expected = [
+            # "manylinux_2_17_i686",  # rejected since it comes before maylinux2014
```
ok

mattip commented on packaging/tags.py:
```diff
@@ -650,28 +681,46 @@ def _linux_platforms(is_32bit=_32_BIT_INTERPRETER):
             linux = "linux_armv7l"
-    manylinux_support = []
     _, arch = linux.split("_", 1)
     if _have_compatible_manylinux_abi(arch):
```
ok

mattip commented:
Requested changes made, there is one test failure with no log.

pradyunsg commented:
Triggered a re-run for all jobs.

pradyunsg commented:
Triggered a "bigger" re-run. :)

mattip commented:
Trying again

mattip commented:
Whack-a-mole. Each time a different CI run fails.

brettcannon commented:
Taking another whack.

pradyunsg commented:
It's as if nicer names made it work better. :P

brettcannon commented on packaging/tags.py:
```diff
@@ -639,7 +670,49 @@ def _have_compatible_manylinux_abi(arch):
+    #
+    # Build a list of maximum glibc versions so that we can
+    # output the canonical list of all glibc from current_glibc
+    # down to too_old_glibc2, including all intermediary versions
```
```suggestion
    # down to too_old_glibc2, including all intermediary versions.
```

mattip replied:
fixing

brettcannon commented on packaging/tags.py:
```diff
@@ -494,9 +513,16 @@ def _glibc_version_string_ctypes():
+
+def _parse_glibc_version(version_str):
+    # type: (str) -> Tuple[int, int]
+    # Parse glibc version
```
```suggestion
    # Parse glibc version.
```

brettcannon commented on packaging/tags.py:
```diff
@@ -639,7 +670,49 @@ def _have_compatible_manylinux_abi(arch):
+
+def _manylinux_tags(linux, arch):
+    # type: (str, str) -> Iterator[str]
+    # Oldest glibc to be supported is (2, 17).
```
```suggestion
    # Oldest glibc to be supported regardless of CPU architecture is (2, 17).
```

mattip replied:
fixing

brettcannon commented:
It might be because it's the end of the day for me, but I can't find a test for https://www.python.org/dev/peps/pep-0600/#package-installers (especially the `manylinux_compatible()` function). Am I just overlooking it?

brettcannon commented:
Otherwise it LGTM as best as I can tell!

mattip commented:
I will add tests (and the needed code) for the https://www.python.org/dev/peps/pep-0600/#package-installers clause in the PEP. I had misssed that.

mattip commented:
Hmm. The current code assumes `once _is_manylinux_compatible() is True, it is True for any lower manylinux tag for this glibc major version` (taken from the comment in the code). As I read the section of the PEP, it seems this is only true when `_manylinux` is not found and default logic is used. Otherwise we must check on every iteration.

Also: the use of `platforms = list(platforms or _platform_tags())` would defeat the iterator character of `_platform_tags` by forcing its evaluation. Is that on purpose?

mattip commented on packaging/tags.py:
```diff
@@ -512,17 +528,9 @@ def _glibc_version_string_ctypes():
-    sys_major, sys_minor = _parse_glibc_version(version_str)
-    return (sys_major, sys_minor) >= (tag_major, tag_minor)
-
-
```
This was merged into the function above

mattip commented on packaging/tags.py:
```diff
@@ -539,20 +547,19 @@ def _parse_glibc_version(version_str):
-    version_str = _glibc_version_string()
-    if version_str is None:
-        return False
-    return _check_glibc_version(version_str, tag_major, tag_minor)
```
`_have_compatible_glibc` did two things, so refactored it out into the component actions

mattip commented on packaging/tags.py:
```diff
@@ -539,20 +547,19 @@ def _parse_glibc_version(version_str):
+        _glibc_version.append((-1, -1))
+    else:
+        _glibc_version.append(_parse_glibc_version(version_str))
+    return _glibc_version[0]
```
use a pseudo-lru_cache to cache the return value from the relatively expensive system call

mattip commented on packaging/tags.py:
```diff
@@ -700,18 +706,12 @@ def _manylinux_tags(linux, arch):
-                )
-                if is_compat:
+                if _is_manylinux_compatible(legacy_tag, arch, glibc_version):
                     yield linux.replace("linux", legacy_tag)
```
We must check `_is_manylinux_compatible` each time, since the user-provided `_manylinux.manylinux_compatible` function is not required to follow any logic

mattip commented on tests/test_tags.py:
```diff
@@ -313,29 +314,16 @@ def test_is_manylinux_compatible_glibc_support(
-        ],
-    )
-    def test_check_glibc_version(self, version_str, major, minor, expected):
-        assert expected == tags._check_glibc_version(version_str, major, minor)
```
What was this test meant to cover? The parsing of the string via the regex? `_check_libc_version` no longer exists. Parsing is tested below

brettcannon replied:
Yes it was.

mattip commented on tests/test_tags.py:
```diff
@@ -454,27 +441,23 @@ def test_linux_platforms_manylinux_unsupported(self, monkeypatch):
+            "manylinux_2_12_" + arch,
             "manylinux2010_" + arch,
             "manylinux_2_11_" + arch,
             "manylinux_2_10_" + arch,
```
Now the expected values include the PEP 600 variant of the legacy tags, which is better.

mattip commented on tests/test_tags.py:
```diff
@@ -601,7 +580,7 @@ def test_linux_platforms_manylinux2014_armv6l(self, monkeypatch):
         self, monkeypatch, machine, abi, alt_machine
     ):
-        monkeypatch.setattr(tags, "_is_manylinux_compatible", lambda name, _: True)
+        monkeypatch.setattr(tags, "_is_manylinux_compatible", lambda name, _: False)
```
This is a weird test. Previously it depended on the monkeypatch on line 42 to set `_have_compatible_glibc` in order to reject any glibc value.

brettcannon replied:
Do you have something specific you want to change that you haven't and so the comment is on its previous state? Or is there something you want to do that you haven't yet?

mattip replied:
Since I refactored `_have_compatible_glibc` the test is more reasonable. Previously `_have_compatible_glibc` did two things, so refactored it out into the component actions did two things, so refactored it out into the component actions

mattip commented:
All but one macOS job passes

mattip commented:
Any thoughts?

brettcannon commented on tests/test_tags.py:
```diff
@@ -285,26 +285,31 @@ def test_mac_platforms(self):
+    @pytest.mark.parametrize(
+        "attribute,glibc", (("1", (2, 5)), ("2010", (2, 12)), ("2014", (2, 17)))
+    )
+    def test_module_declaration(self, manylinux_module, attribute, glibc, tf):
+        manylinux = "manylinux{}_compatible".format(attribute)
+        setattr(manylinux_module, manylinux, tf)
+        res = tags._is_manylinux_compatible(manylinux, "x86_64", glibc)
+        assert tf is res
+        delattr(manylinux_module, manylinux)
```
It's easiest to use pytest's `monkeypatch` for temporarily setting an attribute:

```suggestion
    def test_module_declaration(self, monkeypatch, manylinux_module, attribute, glibc, tf):
        manylinux = "manylinux{}_compatible".format(attribute)
        monkeypatch.setattr(manylinux_module, "manylinux", tf)
        res = tags._is_manylinux_compatible(manylinux, "x86_64", glibc)
        assert tf is res
```

mattip replied:
I tried that. `monkypatch` cannot use `setattr` on an attribute that was not already present, I get
```
>       monkeypatch.setattr(manylinux_module, manylinux, tf)
E       AttributeError: <module '_manylinux'> has no attribute \
                     'manylinux2014_compatible'
```

brettcannon replied:
Try [using `raising=False`](https://docs.pytest.org/en/latest/monkeypatch.html).

mattip commented on packaging/tags.py:
```diff
@@ -509,19 +543,23 @@ def _check_glibc_version(version_str, required_major, minimum_minor):
+    return (int(m.group("major")), int(m.group("minor")))
+
 
+_glibc_version = []  #  type: List[Tuple[int, int]]
```
I needed a `global` variable that I could modify inside `_get_glibc_version` without driving Typing crazy. Is there another quick-and-dirty way to memoize `_get_glibc_version` that is compatible with all the python versions packaging supports?

brettcannon replied:
Until #305 happens, unfortunately not. 😢

mattip commented:
Sorry, I lost the thread of comments that this question relates to:
> Do you have something specific you want to change that you haven't and so the comment is on its previous state? Or is there something you want to do that you haven't yet?

brettcannon commented:
@mattip 
> Sorry, I lost the thread of comments that this question relates to:
>    Do you have something specific you want to change that you haven't and so the comment is on its previous state? Or is there something you want to do that you haven't yet?

It's in response to a comment on `test_tags.py`, line 587 (if you look under the "Files Changed" tab you can find it again):

> This is a weird test. Previously it depended on the monkeypatch on line 42 to set _have_compatible_glibc in order to reject any glibc value.

mattip commented:
Tests are passing with monkeypatch and `raising=False`

brettcannon commented:
@pradyunsg did you want to do another review, or should I go ahead and merge this?

pradyunsg commented:
Go ahead. I'm tied up with college reports and submissions for the next few days.

brettcannon commented:
Thanks for all the work on this, @mattip !
